### ISSUE TITLE
RTCMTLVideoView Build Error

### ISSUE THREAD

**Issue by kenton-hu** [because-why-not/webrtc_visionpro_workspace#4]
<img width="1284" alt="image" src="https://github.com/because-why-not/webrtc_visionpro_workspace/assets/10606844/3f8f860a-5e42-48dd-92bc-2bf5280183ca">

When I add these codes in the MinTestApp, the compiler reported an error, I am unsure of the reason and would appreciate your assistance in reviewing it. Thank you.

**Comment by devluz** [comment]
This file belongs to a component "metal_objc" I had to exclude from the vision os build. This happens within this commit:
https://github.com/because-why-not/webrtc_visionpro/commit/7cf1ae5ea4607593699b24b4e285127c4a175d93#diff-5069c7d69812e0d9e170ed2eec36f1e1b252b3c938a8cb706a9b65c128f5c924R1369

I will check if I can include this

**Comment by kenton-hu** [comment]
<img width="567" alt="image" src="https://github.com/because-why-not/webrtc_visionpro_workspace/assets/10606844/1785df85-c4b2-4ff9-85aa-7121ac53f1c6">

I tried to add it, but the compiler gave an error. Wait for your results.

**Comment by devluz** [comment]
I don't think there is a fix for that. Looks like the point of these classes is to render the video frames as a texture in a MTKView. As there is no MTKView on vision os it is of no use to us.

I think the next best approach is to work with RTCVideoFrame directly and draw it to your own custom UI.

**Comment by youngbody** [comment]
Can I use CAMetalLayer replace MTKView? If it is ok, what should I do base on your code? thanks.

**Comment by devluz** [comment]
You could write your own version of this:
https://github.com/because-why-not/webrtc_visionpro/blob/88f64aabbd0b5c548a1207e28e69a2b0eccc5ec2/sdk/objc/components/renderer/metal/RTCMTLVideoView.m#L11

and then render it to your own custom UI. I don't have any experience with the iOS / visionos UI system so I won't be able to give you much advice for this. My own software uses Unity for UI.

**Comment by magnustviberg** [comment]
I am also getting this issue. Undefined symbol: _OBJC_CLASS_$_RTCMTLVideoView
Is there a fix or workaround?

**Comment by ashok3** [comment]
@devluz would it work if we choose Metal under "Immersive Space Renderer"? 🤔
![Screenshot 2024-09-18 at 2 06 42 PM](https://github.com/user-attachments/assets/41f843a8-b448-4e6f-8298-8830f9bdad4b)

**Comment by devluz** [comment]
@ashok3 Not without manually including files like RTCMTLVideoView.m on your side because I had to remove it. I think the new XCode 16 and vision os SDK 2.0 might include everything we need. I will try a rebuild with these over the next few days and upload it

**Comment by ashok3** [comment]
> @ashok3 Not without manually including files like RTCMTLVideoView.m on your side because I had to remove it. I think the new XCode 16 and vision os SDK 2.0 might include everything we need. I will try a rebuild with these over the next few days and upload it

Yeah, I just found this [answer](https://stackoverflow.com/a/78996178). I'll try on my side, but since you have more ideas about your own implementation, can you please let us know the results anyway?

**Comment by devluz** [comment]
It now builds with the metal dependencies and the visionos SDK 2.0. RTCMTLVideoView should be available now but I haven't tested it. 
see 
https://github.com/because-why-not/webrtc_visionpro_workspace/releases/tag/v0.2.0

Please report back here if it works for you!

**Comment by devluz** [comment]
Was anyone able to check if the issue is resolved with the latest build?

**Comment by cugbwxc** [comment]
> Was anyone able to check if the issue is resolved with the latest build?

It's resolved in  2.0
